concrete/src/Package/Package.php

```diff
@@ -367,205 +580,279 @@ public function uninstall()
+        } elseif ($this->getPackageHandle() === '' || !is_dir($this->getPackagePath())) {
             $errors[] = self::E_PACKAGE_NOT_FOUND;
         }
 
         // Step 2 - check to see if the user has already installed a package w/this handle
         if ($testForAlreadyInstalled) {
             $entity = $this->getPackageEntity();
-            if (is_object($entity) && $entity->isPackageInstalled()) {
+            if ($entity !== null && $entity->isPackageInstalled()) {
                 $errors[] = self::E_PACKAGE_INSTALLED;
             }
         }
 
-        if (count($errors) == 0) {
-            // test minimum application version requirement
-            if (version_compare(APP_VERSION, $this->getApplicationVersionRequired(), '<')) {
-                $errors[] = [self::E_PACKAGE_VERSION, $this->getApplicationVersionRequired()];
+        if (empty($errors)) {
+            // Step 3 - test minimum application version requirement
+            $applicationVersionRequired = $this->getApplicationVersionRequired();
+            if (version_compare(APP_VERSION, $applicationVersionRequired, '<')) {
+                $errors[] = [self::E_PACKAGE_VERSION, $applicationVersionRequired];
             }
-        }
 
-        if (count($errors) > 0) {
-            $e = $this->app->make('error');
-            foreach ($errors as $error) {
-                $e->add($this->getErrorText($error));
+            // Step 4 - Check for package dependencies
+            $packageService = $this->app->make(PackageService::class);
+            /* @var PackageService $packageService */
+            $installedPackageHandles = $packageService->getInstalledHandles();
+            $myDependencies = $this->getPackageDependencies();
+            foreach ($myDependencies as $packageHandle => $requirements) {
+                if (in_array($packageHandle, $installedPackageHandles)) {
+                    if ($requirements === false) {
+                        $errors[] = t('This package can\'t be installed if the package with handle %s is already installed', $packageHandle);
+                    } elseif ($requirements !== true) {
+                        $packageVersion = $packageService->getByHandle($packageHandle)->getPackageVersion();
+                        if (is_array($requirements)) {
+                            if (version_compare($packageVersion, $requirements[0]) < 0 || version_compare($packageVersion, $requirements[1]) > 0) {
+                                $errors[] = t('This package requires the package with handle %1$s has a version from %2%s to %3$s', $packageHandle, $requirements[0], $requirements[1]);
+                            }
+                        } else {
+                            if (version_compare($packageVersion, $requirements) < 0) {
+                                $errors[] = t('This package requires the package with handle %1$s has a version %2%s or greater', $packageHandle, $requirements);
```

Comment:
'This package requires `that` package with handle %1$s has a version %2%s or greater'